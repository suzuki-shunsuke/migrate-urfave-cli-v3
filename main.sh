#!/usr/bin/env bash

set -euxo pipefail

patterns=(
	's|"github\.com/urfave/cli/v2"|"github.com/urfave/cli/v3"|'
	's|"github\.com/urfave/cli/v2/altsrc"|"github.com/urfave/cli-altsrc/v3"|'
	's|cli\.App|cli.Command|'
	's|EnableBashCompletion|EnableShellCompletion|'
	's|RunContext|Run|'
	's|Subcommands|Commands|'
	's|CustomAppHelpTemplate|CustomRootCommandHelpTemplate|'
	's|cli\.NewApp\(\)|\&cli.Command{}|'
	's|\(([^ ]+?) \*cli\.Context\) error|(ctx context.Context, \1 *cli.Command) error|'
	's|\(\*cli\.Context\) error|(context.Context, *cli.Command) error|'
	's|ExitErrHandler = func\(\*cli\.Context, error\)|ExitErrHandler = func(context\.Context, *cli.Command, error)|'
	's|\*cli\.Context|context.Context|'
	's|EnvVars: \[\]string\{([^}]+)\}|Sources: cli.EnvVars(\1)|'
	's|EnvVars: \[\]string|Sources: cli.EnvVars|'
)

while read -r file; do
	for p in "${patterns[@]}"; do
		sed -Ei "$p" "$file"
	done
	sed -Ei "/^	*Compiled: /d" "$file"
done < <(git ls-files | grep -E '\.go$')

go mod tidy
