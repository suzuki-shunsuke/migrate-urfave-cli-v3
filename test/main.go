package main

import (
	"context"
	"fmt"
	"os"

	"github.com/suzuki-shunsuke/urfave-cli-help-all/helpall"
	"github.com/urfave/cli/v2"
)

func main() {
	_ = cli.NewApp()
	app := &cli.App{
		Name: "example",
		Authors: []*cli.Author{
			{
				Name:  "name",
				Email: "email",
			},
		},
		EnableBashCompletion:  true,
		CustomAppHelpTemplate: "{{.Name}} - {{.Usage}}",
		ExitErrHandler: func(_ *cli.Context, err error) {
			if err != nil {
				os.Exit(1)
			}
		},
		Commands: []*cli.Command{
			{
				Name: "sub",
				Subcommands: []*cli.Command{
					{
						Name: "subsub",
						Action: func(c *cli.Context) error {
							name := c.String("name")
							age := c.Int("age")
							c.App.ToFishCompletion()
							fmt.Fprintln(os.Stdout, name, age)
							return nil
						},
						Flags: []cli.Flag{
							&cli.StringFlag{
								Name:    "name",
								EnvVars: []string{"NAME"},
							},
						},
					},
				},
			},
			helpall.New(nil),
		},
	}
	app.RunContext(context.Background(), os.Args)
}
