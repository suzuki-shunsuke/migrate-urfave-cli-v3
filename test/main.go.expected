package main

import (
	"context"
	"fmt"
	"os"

	"github.com/suzuki-shunsuke/urfave-cli-v3-help-all/helpall"
	"github.com/urfave/cli/v3"
)

func main() {
	_ = &cli.Command{}
	app := &cli.Command{
		Name: "example",
		Authors: []any{
			{
				Name:  "name",
				Email: "email",
			},
		},
		EnableShellCompletion:  true,
		CustomRootCommandHelpTemplate: "{{.Name}} - {{.Usage}}",
		ExitErrHandler: func(_ context.Context, err error) {
			if err != nil {
				os.Exit(1)
			}
		},
		Commands: []*cli.Command{
			{
				Name: "sub",
				Commands: []*cli.Command{
					{
						Name: "subsub",
						Action: func(ctx context.Context, c *cli.Command) error {
							name := c.String("name")
							age := c.Int("age")
							c.ToFishCompletion()
							fmt.Fprintln(os.Stdout, name, age)
							return nil
						},
						Flags: []cli.Flag{
							&cli.StringFlag{
								Name:    "name",
								Sources: cli.EnvVars("NAME"),
							},
						},
					},
				},
			},
			helpall.New(nil),
		},
	}
	app.Run(context.Background(), os.Args)
}
